### DISCUSSION TITLE
Is there any performance difference between `#?#:-abp-has()` and `#?#:has()` on browsers that support the has selector?

### DISCUSSION THREAD

**Discussion by partingscientist** [uBlockOrigin/uAssets#21166] [Q&A] [answered]
With the release of Firefox 121 in less that two weeks (December 19th) along with the `has()` selector, is there any performance difference between `#?#:-abp-has()` and `#?#:has()` in Firefox for uBO? I have previously read that uBO discards the `?` if I'm not mistaken, which I would assume mean that `#?#:has()` is handled just like native has selector if available. Is this correct? Is `#?#:-abp-has()` also converted to native has if the browser supports it? Thank you.

**Comment by krystian3w** [comment]
It should be same like on Chrome 105+ => can be used as CSS 4.

It looks CSSTree implementation made simple `:has()` filtering non-procedural (by uBo implementation). 

You can check on Beta/Nightly in how method was forced (JS with random attributes or CSS 4).

---

As for the hiding speed, the blinking is very minimal (the only limitation is CSS and browser logic now).

**Comment by garry-ut99** [comment]
In theory a native `:has` is supposed to be faster than a procedural `:has`, in practice I didn't test speed so far, so I dunno.

Whether some of uBO procedural filters are handled as native selectors is another story, some info is in this thread : https://github.com/uBlockOrigin/uBlock-issues/issues/2475, I know that `#?#:-abp-has()` is an alias of `##:has()` which is converted to native on browsers with native support, and adding `?` ==> `#?#:has()` aborts the conversion, and forces to remain as a procedural, I'm not sure whether anything changed since then.

**Reply by krystian3w** [reply]
Now `#?#` can't allow back to JS implementation of hiding.

**Reply by partingscientist** [reply]
> `#?#:has()` aborts the conversion, and forces to remain as a procedural

@garry-ut99 So is it true that all `#?#:has()` filters in Adguard list do not use the native selector when used in uBO? That does not sound correct.

**Reply by krystian3w** [reply]
No longer aborts.

Now needed is Invalid CSS4 selector or procedular modifier (e.g. `has-text`).

Example something Looks as CSS4 but is not:

```adb
##:not(.item *):not(:has(.item *))

! uBo parse same to:
! ##.item:others()
```

**Reply by gorhill** [reply, marked as answer]
`#?#:-abp-has()` is equivalent to `##:has()`. ~~`#?#:has()` will force the filter to be procedural, this is to ensure there is a fallback syntax should the result be unexpected when using the native form.~~ Never mind, `#?#:has()` is also compiled to native if supported by the browser.

**Reply by partingscientist** [reply]
I originally asked this because I thought migrating from `#?#:-abp-has()` to `#?#:has()` will use the native selector instead of procedural in uBO (considering ABP supports `#?#:has()` as an alias to `#?#:-abp-has()`).

Now I'm curious, does `#?#:has()` filters in built-in Adguard filters are treated as being procedural instead of native?

**Reply by krystian3w** [reply]
They also start mostly as CSS 4, check with DevTools that most of working ads/annoy&shy;ances them do not have 2 attributes added with a random name.

In general, it looks like Microsoft implemented a shadowban on repositories or comments sections.

**Reply by gorhill** [reply]
> does #?#:has() filters in built-in Adguard filters are treated as being procedural

~~Yes, this forces procedural in uBO. You would have to ask them why they use `#?#` instead of `##`.~~ I am wrong, see below.

**Reply by krystian3w** [reply]
So why this start in CSS4:

```adb
example.*#?#div:has(a)
```

**Reply by gorhill** [reply]
You're right, `#?#` is compiled to native. I had looked at the code and misread [that part ](https://github.com/gorhill/uBlock/blob/bda46d5cc7fb0994a8b4a28046d78832d140eb64/src/js/static-filtering-parser.js#L3259).

**Reply by u-RraaLL** [reply]
So the only times `:has()` will be a procedural now is when used with a natively unsupported subject or argument?

**Reply by gorhill** [reply]
Yes.

**Comment by u-RraaLL** [comment]
>> So the only times `:has()` will be a procedural now is when used with a natively unsupported subject or argument?

> Yes.

Does `:matches-path()` also degrade `:has()` to a procedural one? Or is it implemented in a way that has no bearing on that? @gorhill 

I see a twitch filter in experimental using both and was wondering if it should be switched to `:upward()` instead. Or maybe move `:matches-path()` after the selectors?

**Reply by uBlock-user** [reply]
You mean `www.twitch.tv##:matches-path(/^\/[0-9_a-z]+$/) [data-a-target="video-ref"]:has(.video-player__overlay > .follow-panel-overlay.tw-transition[aria-hidden="false"]) + div[class^="Layout-sc-"][hidden=""]:style(display:block !important;)` ? I don't think so. The entire `:has()` has supported arguments.  Also keep in mind `:upward()` has created false positives unlike `:has()`, where arguments are specific, so better to avoid that.

**Reply by u-RraaLL** [reply]
I mean that `:matches-path()` is a procedural or at least is listed as one. And using `:has()` after procedurals makes it procedural too.

It could also be that it's the only "safe" procedural in this regard. I don't know, which is why I'm asking.,

It technically works anywhere in the filter too, so maybe pushing it to the end would be enough to fix the issue, if it exists at all.

**Reply by u-RraaLL** [reply]
Okay, it should be safe. If I'm not mistaken, this would've failed to filter if `:has()` was procedural.

    github.com##:matches-path(uBlockOrigin/uAssets/discussions) li.prc-UnderlineNav-UnderlineNavItem-syRjR:has(>[href="/uBlockOrigin/uAssets/pulse"],>[href="/uBlockOrigin/uAssets/projects"])

**Reply by gorhill** [reply]
> is it implemented in a way that has no bearing on that?

Yes. The only time `:has(...)` is compiled to procedural one is when the argument of `:has(...)` is not a valid one as per CSS4 specs -- i.e. `...:has(a:has-text(...))`.

**Reply by u-RraaLL** [reply]
I thought that both subject and argument need to be valid CSS.

Doesn't `a:has-text(b):has(c)` compile to a procedural `:has()`? So I was wondering if `matches-path()` is considered part of the subject.

Basically, I thought that `a:procedural(b):has(c)` makes `:has()` procedural too, while `a:has(b):procedural(c)` leaves `:has()` native, so a good way to narrow down nodes before using a procedural. It that wrong?

**Reply by krystian3w** [reply]
You should use `:has(> :is([href="/uBlockOrigin/uAssets/pulse"], [href="/uBlockOrigin/uAssets/projects"]))`

**Reply by gorhill** [reply]
> Doesn't `a:has-text(b):has(c)` compile to a procedural `:has()`?

No, it will be native `:has()`. The issue though with that filter is that for the part following `:has-text()`, there is a need to convert the found elements to `:nth-child()` such that the resulting selector is `:scope > :nth-child([position of matching element in parent]):has(c)`.
